### PR TITLE
implement non-controversial cleanup portions of #18762

### PR DESCRIPTION
**Pull request by jbrockmendel**
#18762 touches a lot of code and is currently in limbo pending discussion of `roll_monthday` and `roll_convention`.  This separates out unrelated cleanup portions of that PR, the idea being that making the diff there simpler will make things easier.

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/_libs/tslibs/offsets.pyx [hunk 1]
```diff
@@ -836,17 +833,20 @@ cpdef int roll_yearday(other, n, month, day_opt='start') except? -1:
     ----------
     other : datetime or Timestamp
     n : number of periods to increment, before adjusting for rolling
-    day_opt : 'start', 'end'
-        'start': returns 1
-        'end': returns last day  of the month
+    month : reference month giving the first month of the year
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: I would make this a required parameter and pass it. its more obvious.

**Review by jreback — CHANGES_REQUESTED** [review]
may day_opt in shift_month required as well. optional args to these routines are quite error prone.

**Comment by jreback** [comment]
looks fine. do we need additional test coverage of shift_months, roll_yearday to account for not having None's? (I think yes, if nothing broke with this change)

**Comment by jbrockmendel** [comment]
> do we need additional test coverage of shift_months, roll_yearday to account for not having None's? (I think yes, if nothing broke with this change)

I'm fine with adding that to the todo list, but don't want this PR to descend into yak shaving.

File: pandas/_libs/tslibs/offsets.pyx [hunk 2]
```diff
@@ -587,15 +583,17 @@ def shift_quarters(int64_t[:] dtindex, int quarters,
...
 @cython.wraparound(False)
 @cython.boundscheck(False)
-def shift_months(int64_t[:] dtindex, int months, object day=None):
+def shift_months(int64_t[:] dtindex, int months, object day):
     """
     Given an int64-based datetime index, shift all elements
     specified number of months using DateOffset semantics
 
-    day: {None, 'start', 'end'}
+    day: {None, 'start', 'end', 'business_start', 'business_end'}
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: remove the None as a valid argument or is it? also need to fix the code

**Review by jreback — CHANGES_REQUESTED** [review]
this is not yak shaving. believe me you would know that! rather putting fixes in a single PR is better for review.

File: pandas/_libs/tslibs/offsets.pyx (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/offsets.pyx]: None is valid.

**Comment by jreback** [comment]
superseded by #18762

**Pull request closed without merging**